AnkiDroid/src/main/java/com/ichi2/anki/instantnoteeditor/InstantEditorViewModel.kt

```diff
@@ -0,0 +1,192 @@
+    private val oldNoteId = MutableSharedFlow<Long?>()
+
+    private val _editorNote = MutableLiveData<Note?>()
+    val editorNote: LiveData<Note?> get() = _editorNote
+
+    private val _currentlySelectedNotetype = MutableLiveData<NotetypeJson>()
+    val currentlySelectedNotetype: LiveData<NotetypeJson> get() = _currentlySelectedNotetype
+
+    private val _deckId = MutableLiveData<DeckId>(0)
+    val deckId: LiveData<DeckId> get() = _deckId
+
+    /** Doesn't finish the activity when the editor dialog is dismissed on save note,
+     * thus allowing snackbar to be visible to the user */
+    private val _savingNote = MutableLiveData(false)
+    val savingNote: LiveData<Boolean> get() = _savingNote
+
+    init {
+        viewModelScope.launch {
+            oldNoteId.emit(null)
+        }
+    }
+
+    fun setDeckId(deckId: DeckId) {
+        _deckId.value = deckId
+    }
+
+    fun setSavingNote(value: Boolean) {
+        _savingNote.value = value
+    }
+
+    suspend fun getCurrentDeck() {
+        val did = withCol { config.get(Decks.CURRENT_DECK) ?: 1L }
+        _deckId.postValue(did)
+    }
+
+    private suspend fun changeToClozeType() {
+        Timber.d("Changing to cloze type note")
+        val model = withCol {
+            notetypes.all().first { noteType ->
+                noteType.isCloze
+            }
+        }
+
+        _currentlySelectedNotetype.postValue(model)
+
+        val oldModelId = withCol { notetypes.current().getLong("id") }
+
+        Timber.i("Changing note type to '%d", model.id)
+        viewModelScope.launch {
+            oldNoteId.emit(oldModelId)
+        }
+
+        withCol {
+            setModel(this, model.id)
+        }
+    }
+
+    fun checkAndSaveNote(activity: Activity) {
+        viewModelScope.launch {
+            val noteFieldsCheck = editorNote.value?.let { checkNoteFieldsResponse(it) }
+            if (noteFieldsCheck is NoteFieldsCheckResult.Failure) {
+                // TODO: check if we need more errors here like note editor
+                instantEditorError.emit(noteFieldsCheck.getLocalizedMessage(activity))
+                setSavingNote(false)
+                return@launch
+            }
+            Timber.d("Note fields check successful, saving note")
+            instantEditorError.emit(null)
+            saveWithProgress(activity)
+        }
+    }
+
+    suspend fun saveWithProgress(activity: Activity) {
+        putDeckIdInNote()
+        activity.withProgress(activity.resources.getString(R.string.saving_facts)) {
+            val note = editorNote.value ?: return@withProgress
+            val deckId = deckId.value ?: return@withProgress
+            undoableOp { addNote(note, deckId) }
+        }
+        noteSaved.emit(true)
+    }
+
+    suspend fun updateModel() {
+        val nid = oldNoteId.first()
+        withCol {
+            if (nid != null) {
+                setModel(this, nid)
+            }
+        }
```

Comment:
If you can move a check outside the `withCol`, do so, as this saves locking the collection

Combined with the above comment

```suggestion
        val nid = oldNoteId.firstOrNull() ?: return
        withCol { setModel(this, nid) }
```